### PR TITLE
[Task #4] Add controllers to API project

### LINKED ISSUE #9: Add Controllers to API proj
Create BaseController
Add FluentValidation (https://docs.fluentvalidation.net/en/latest/installation.html)

### LINKED ISSUE #11: Add Swagger to project
https://docs.microsoft.com/en-us/aspnet/core/tutorials/getting-started-with-swashbuckle?view=aspnetcore-5.0&tabs=visual-studio

### LINKED ISSUE #12: Implement ErrorHandler middleware  and Write custom Exception(ItemNotFoundException)
https://code-maze.com/global-error-handling-aspnetcore/

### PR DESCRIPTION
**Pull request by Aleksander-Drumev-WA**
Controllers created.
FluentValidations added to project.
Web project became web API.

### CODE REVIEW

File: WA.Pizza.Web/Controllers/BaseController.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using Microsoft.AspNetCore.Mvc;
+
+namespace WA.Pizza.Web.Controllers
+{
+    [ApiController]
+    [Route("api/[controller]")]
```
Comment (var1ap-WA) [inline WA.Pizza.Web/Controllers/BaseController.cs:7]: Add Authorize attribute here

File: WA.Pizza.Web/Controllers/OrdersController.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
+﻿using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using WA.Pizza.Infrastructure.Data.Services;
+
+namespace WA.Pizza.Web.Controllers
+{
+    [Authorize]
```
Comment (var1ap-WA) [inline WA.Pizza.Web/Controllers/OrdersController.cs]: remove here

File: WA.Pizza.Web/Controllers/OrdersController.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        private readonly OrderDataService _orderDataService;
+
+        public OrdersController(OrderDataService orderDataService)
+        {
+            _orderDataService = orderDataService;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> CreateOrder(int basketId)
```
Comment (var1ap-WA) [inline WA.Pizza.Web/Controllers/OrdersController.cs]: IActionResult is not necessary

File: WA.Pizza.Web/Controllers/BaseController.cs (hunk 1 above)
Comment (Aleksander-Drumev-WA) [reply WA.Pizza.Web/Controllers/BaseController.cs:7]: Done

File: WA.Pizza.Web/Controllers/OrdersController.cs (hunk 2 above)
Comment (Aleksander-Drumev-WA) [reply WA.Pizza.Web/Controllers/OrdersController.cs]: Removed

File: WA.Pizza.Web/Controllers/OrdersController.cs (hunk 3 above)
Comment (Aleksander-Drumev-WA) [reply WA.Pizza.Web/Controllers/OrdersController.cs]: Removed everywhere

File: Pizzeria.Tests/BasketDataServiceTests.cs [hunk 4]
```diff
@@ -222,14 +222,66 @@ public async Task User_changes_basketItem_quantity()
...
 
             // Assert
             var updatedCatalogItem = await _dbContext.BasketItems.FindAsync(updatedItemId);
-            updatedCatalogItem.Quantity.Should().Be(22);
+            updatedCatalogItem!.Quantity.Should().Be(22);
+            updatedCatalogItem.CatalogItem.StorageQuantity.Should().Be(resultStorageQuantity);
+        }
+
+        [Fact]
+        public async Task User_wants_more_quantity_than_storage_quantity_when_updating_basket_item()
```
Comment (var1ap-WA) [inline Pizzeria.Tests/BasketDataServiceTests.cs]: too long and not really correct
`user try to order more than  we have on storage`

File: WA.Pizza.Web/Controllers/OrdersController.cs [hunk 5]
```diff
@@ -32,6 +32,7 @@ public Task UpdateOrder(int userId, string orderStatus)
         }
 
         [HttpGet]
+        [Route("order")]
```
Comment (var1ap-WA) [inline WA.Pizza.Web/Controllers/OrdersController.cs]: you can add route with [HttpGet]

File: WA.Pizza.Web/Program.cs [hunk 6]
```diff
@@ -37,8 +39,15 @@
...
 
+app.CustomExceptionHandlerMiddleware();
 app.UseHttpsRedirection();
-app.UseStaticFiles();
+
+app.UseSwagger();
+app.UseSwaggerUI(s =>
+{
+    s.SwaggerEndpoint("/swagger/v1/swagger.json", "My API V1");
+    s.RoutePrefix = string.Empty;
```
Comment (var1ap-WA) [inline WA.Pizza.Web/Program.cs]: better use `/swagger` prefix

File: Pizzeria.Tests/BasketDataServiceTests.cs (hunk 4 above)
Comment (Aleksander-Drumev-WA) [reply Pizzeria.Tests/BasketDataServiceTests.cs]: Renamed

File: WA.Pizza.Web/Controllers/OrdersController.cs (hunk 5 above)
Comment (Aleksander-Drumev-WA) [reply WA.Pizza.Web/Controllers/OrdersController.cs]: Done

File: WA.Pizza.Web/Program.cs (hunk 6 above)
Comment (Aleksander-Drumev-WA) [reply WA.Pizza.Web/Program.cs]: Done.

File: WA.Pizza.Web/BackgroundJobs/ErrorHandlerMiddleware.cs [hunk 7]
```diff
@@ -30,13 +32,26 @@ public async Task InvokeAsync(HttpContext httpContext)
 
         private async Task HandleExceptionAsync(HttpContext context, Exception exception)
         {
+            var pathFeature = context.Features.Get<IExceptionHandlerPathFeature>();
             context.Response.ContentType = "application/json";
-            context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
-            await context.Response.WriteAsync(new ErrorDetails()
+            if (exception.GetType().FullName == "ItemNotFoundException")
+            {
+                context.Response.StatusCode = (int)HttpStatusCode.NotFound;
```
Comment (var1ap-WA) [inline WA.Pizza.Web/BackgroundJobs/ErrorHandlerMiddleware.cs]: StatusCodes.Status404NotFound

Comment (Aleksander-Drumev-WA) [reply WA.Pizza.Web/BackgroundJobs/ErrorHandlerMiddleware.cs]: Done

**Review by var1ap-WA — APPROVED** [review]

**Pull request closed without merging**